brainglobe_workflows/cellfinder/cellfinder_main.py

```diff
@@ -0,0 +1,305 @@
+    model_weights: Optional[os.PathLike]
+    model: str
+    batch_size: int
+    n_free_cpus: int
+    network_voxel_sizes: Tuple[int, int, int]
+    soma_diameter: int
+    ball_xy_size: int
+    ball_z_size: int
+    ball_overlap_fraction: float
+    log_sigma_size: float
+    n_sds_above_mean_thresh: int
+    soma_spread_factor: float
+    max_cluster_size: int
+    cube_width: int
+    cube_height: int
+    cube_depth: int
+    network_depth: depth_type
+
+    list_signal_files: Optional[list] = None
+    list_background_files: Optional[list] = None
+
+
+def example_cellfinder_script():
+    cfg = setup_workflow()  # (this won't be timed)
+    run_workflow_from_cellfinder_run(cfg)  # (this will be timed)
+
+
+def run_workflow_from_cellfinder_run(cfg):
+    """
+    Run workflow based on the cellfinder_core.main.main()
+    function.
+
+    The steps are:
+    1. Read the input signal and background data as two separate
+       Dask arrays.
+    2. Run the main cellfinder pipeline on the input Dask arrays,
+       with the parameters defined in the input configuration (cfg).
+    3. Save the detected cells as an xml file to the location specified in
+       the input configuration (cfg).
+
+    We plan to time each of the steps in the workflow individually,
+    as well as the full workflow.
+
+    Parameters
+    ----------
+    cfg : CellfinderConfig
+        a class with the required setup methods and parameters for
+        the cellfinder workflow
+    """
+    # Read input data as Dask arrays
+    signal_array = read_with_dask(cfg.signal_parent_dir)
+    background_array = read_with_dask(cfg.background_parent_dir)
+
+    # Run main analysis using `cellfinder_run`
+    detected_cells = cellfinder_run(
+        signal_array, background_array, cfg.voxel_sizes
+    )
+
+    # Save results to xml file
+    save_cells(detected_cells, cfg.detected_cells_filepath)
+
+
+def setup_workflow(default_config_dict: dict = default_config_dict):
+    """Prepare configuration to run workflow
+
+    This includes
+    - instantiating the config dictionary,
+    - checking if the input data exists locally, and fetching from
+      GIN repository otherwise,
+    - creating the directory for the output of the workflow if it doesn't exist
+
+    To instantiate the config dictionary, we first check if an environment
+    variable "CELLFINDER_CONFIG_PATH" pointing to a config json file exists.
+    If not, the default config is used.
+
+    Parameters
+    ----------
+    default_config_dict : dict
+        a dictionary with the default config parameters
+
+    Returns
+    -------
+    _type_
+        _description_
+    """
+
+    # Define config
+    if "CELLFINDER_CONFIG_PATH" in os.environ.keys():
+        input_config_path = Path(os.environ["CELLFINDER_CONFIG_PATH"])
+        assert input_config_path.exists()
+
+        # read into dict (assuming config is a json dict?)
+        # TODO:add error handling here?
```

Comment:
This is a good point: we might want to move from configuration dataclasses to `pydantic` or `attrs` for implicit validation? Open an issue?